src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs

```diff
@@ -171,25 +172,41 @@ public virtual SecurityToken CreateToken(SecurityTokenDescriptor tokenDescriptor
         }
 
         /// <inheritdoc/>
-        public override Task<TokenValidationResult> ValidateTokenAsync(string token, TokenValidationParameters validationParameters)
+        public override async Task<TokenValidationResult> ValidateTokenAsync(string token, TokenValidationParameters validationParameters)
         {
             try
             {
-                var claimsPrincipal = ValidateToken(token, validationParameters, out var validatedToken);
-                return Task.FromResult(new TokenValidationResult
+                if (string.IsNullOrEmpty(token))
+                    throw LogArgumentNullException(nameof(token));
+
+                if (validationParameters == null)
+                    throw LogArgumentNullException(nameof(validationParameters));
+
+                if (token.Length > MaximumTokenSizeInBytes)
+                    throw LogExceptionMessage(new ArgumentException(FormatInvariant(TokenLogMessages.IDX10209, LogHelper.MarkAsNonPII(token.Length), LogHelper.MarkAsNonPII(MaximumTokenSizeInBytes))));
+
+                validationParameters = await PopulateValidationParametersWithCurrentConfigurationAsync(validationParameters).ConfigureAwait(false);
+
+                var samlToken = ValidateSignature(token, validationParameters);
+                if (samlToken == null)
+                    throw LogExceptionMessage(
+                        new SecurityTokenValidationException(FormatInvariant(TokenLogMessages.IDX10254, LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII("ValidateToken"), LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII("ValidateSignature"), LogHelper.MarkAsNonPII(typeof(Saml2SecurityToken)))));
+                var claimsPrincipal = ValidateToken(samlToken, token, validationParameters, out var validatedToken);
```

Comment:
ValidateToken calls ValidateSignature.

It may be clearer to have ValidateToken call ValidateTokenAsync and have all logic in ValidateTokenAsync.

Reply:
I was a little hesitant about doing that because of the possible deadlock situation that can arise from calling async code from sync context. I realise that I already do that with PopulateValidationParametersWithCurrentConfigurationAsync so there should not be any difference in that regard.
I will make the change.

Reply:
I have tried implementing this, but I'm strugling to find a neat way that preserves stack trace information on exceptions.

Reply:
> ValidateToken calls ValidateSignature.

I tried to understand what you meant by this. That is the way to do it right? Validate signature to get the Saml2SecurityToken object